### PR TITLE
Validate Cookie

### PR DESCRIPTION
**Pull request by ndt93**
Implement cookie validation for every API request using both in memory cache and database access fallback for faster response time and reliability.

### CODE REVIEW

**Comment by ndt93** [comment]
@yj-soh Ready for review

File: app_server/.gitignore [hunk 1]
```diff
@@ -29,3 +29,4 @@ build/Release
 node_modules
 doc
 report
+.tern-project
```
Comment (yj-soh) [inline app_server/.gitignore]: Use your local `exclude` file instead if it's not relevant to the all contributors.

File: app_server/app/Router.js [hunk 2]
```diff
@@ -63,6 +55,39 @@ server.register({
   }
 });
 
+/* Configure Authentication plugin */
+server.register(require('hapi-auth-cookie'), function (err) {
+  var cache = server.cache({segment: 'sessions',
+                            expiresIn: 3 * 24 * 60 * 60 * 1000});
+  server.app.cache = cache;
```
Comment (yj-soh) [inline app_server/app/Router.js]: Minor issue, but is there a reason for the intermediate `cache` variable here?

File: app_server/app/controllers/UserController.js [hunk 3]
```diff
@@ -69,10 +95,21 @@ Class.logout = function (request, reply) {
   reply('Logged out!');
 };
 
+function clearUserPrivateInfo(user) {
+  if (user.password) {
+    delete user.password;
+  }
+  if (user.accessToken) {
+    delete user.accessToken;
+  }
```
Comment (yj-soh) [inline app_server/app/controllers/UserController.js]: I'm not entirely sure on this either, but is there any purpose in checking if the properties exist before deleting them?
I tried testing a little, but the additional check seems to be irrelevant.
(it may cause trouble in strict mode if `user` isn't an object, but that would happen here too)

File: app_server/app/policies/Authenticator.js [hunk 4]
```diff
@@ -14,11 +14,14 @@ var Utility = rfr('app/util/Utility');
 var logger = Utility.createLogger(__filename);
 
 function Authenticator() {
+  this.SID = 'sid-worldscope';
 }
 var Class = Authenticator.prototype;
 
 Class.ERRORS = {
   RETRIEVE_PROFILE: 'Error retrieving user\'s social media profile',
+  INVALID_CREDENTIALS: 'username or password is invalid',
```
Comment (yj-soh) [inline app_server/app/policies/Authenticator.js]: Should this be capitalized?

File: app_server/app/util/Utility.js [hunk 5]
```diff
@@ -44,7 +44,8 @@ exports.createLogger = function (filename) {
       new winston.transports.Console({
         timestamp: true,
         colorize: true,
-        label: getModuleName(filename)
+        label: getModuleName(filename),
+        level: process.env.NODE_ENV == 'development' ? 'debug' : 'info'
```
Comment (yj-soh) [inline app_server/app/util/Utility.js]: `===`?
(`TestUtils.js` too.)

File: app_server/test/controllers/UserController.js [hunk 6]
```diff
@@ -3,30 +3,59 @@ var Lab = require('lab');
...
   lab.test('Get list of users', function (done) {
-    Router.inject('/api/users/asd', function (res) {
-      var errorMsg = 'child "id" fails because ["id" must be a number]';
-      Code.expect(res.result.statusCode).to.equal(400);
-      Code.expect(res.result.message).to.equal(errorMsg);
-      done();
-    });
+    Router.inject({url: '/api/users/asd', credentials: testAccount},
+      function (res) {
+        var errorMsg = 'child "id" fails because ["id" must be a valid GUID]';
```
Comment (yj-soh) [inline app_server/test/controllers/UserController.js]: The name of this test should probably be "Invalid id".
Maybe throw in some symbols instead of "asd"?
And what's a child id?

File: app_server/test/controllers/UserController.js [hunk 7]
```diff
@@ -83,8 +125,48 @@ lab.experiment('UserController log in tests', function () {
...
+
+  lab.test('Valid login', function (done) {
+    facebookServer.start(function (err) {
+      Code.expect(err).to.not.be.null();
+
+      Facebook.FACEBOOK_API_URL = 'http://localhost:8888';
+
+      Router.inject({method: 'POST', url: '/api/users/login',
+                    payload: {appId: '123456789',
+                              accessToken: 'xyz'}}, function (res) {
```
Comment (yj-soh) [inline app_server/test/controllers/UserController.js]: You probably intended to add another space before `payload` (and subsequently `accessToken`) here?
(likewise in `test/policies/Authenticator.js`)

File: app_server/test/controllers/UserController.js [hunk 8]
```diff
@@ -83,8 +125,48 @@ lab.experiment('UserController log in tests', function () {
...
+          platformId: '18292522986296117'
+        };
+
+        Code.expect(res.statusCode).to.equal(200);
+
+        var body = JSON.parse(res.payload);
+        Code.expect(body.alias).to.equal(expected.alias);
+        Code.expect(body.username).to.equal(expected.username);
+        Code.expect(body.platformType).to.equal(expected.platformType);
+        Code.expect(body.platformId).to.equal(expected.platformId);
```
Comment (yj-soh) [inline app_server/test/controllers/UserController.js]: Will similar segments be used in the future?
Should we extract this into some function (in TestUtils?) that takes in 2 objects and checks if all of the properties in one are equal to the relevant properties in the other?

File: app_server/test/controllers/UserController.js [hunk 9]
```diff
@@ -83,8 +125,48 @@ lab.experiment('UserController log in tests', function () {
...
+
+        Facebook.FACEBOOK_API_URL = 'https://graph.facebook.com';
+        facebookServer.stop(function () {});
```
Comment (yj-soh) [inline app_server/test/controllers/UserController.js]: If this is a callback, is there supposed to be something happening there?

File: app_server/test/services/UserService.js [hunk 10]
```diff
@@ -85,6 +85,24 @@ lab.experiment('UserService tests', function () {
...
+    UserService.createNewUser(bob).then(function (result) {
+      return UserService.getUserById(result.userId);
+    }).then(function(user) {
+      Code.expect(user.username).to.equal(bob.username);
+      Code.expect(user.password).to.equal(bob.password);
+      done();
+    });
+  });
+
+  lab.test('getUserById invalid arguments', function(done) {
```
Comment (yj-soh) [inline app_server/test/services/UserService.js]: Should we have these tests both here and in `Service`?
As far as I can tell, it shouldn't improve either stability nor coverage as opposed to testing in `Service` only...?

File: app_server/app/Router.js (hunk 2 above)
Comment (ndt93) [inline app_server/app/Router.js]: Ok, can be removed. It was previously used in validate code below.

File: app_server/app/controllers/UserController.js (hunk 3 above)
Comment (ndt93) [inline app_server/app/controllers/UserController.js]: Ok sure. Didn't know `delete` is that smart.

File: app_server/test/controllers/UserController.js (hunk 9 above)
Comment (ndt93) [inline app_server/test/controllers/UserController.js]: Supposed to check for error there but doesn't matter in testing context here since the process will be killed off anyway after the test is ran.

File: app_server/test/services/UserService.js (hunk 10 above)
Comment (ndt93) [inline app_server/test/services/UserService.js]: Ideally, methods in `Service`should be tested using verification (verify that it calls `UserService` once and receive corresponding result) so that it doesn't need to run through the whole flow but still can confirm the link between `Service` and `UserService` . But that's quite messy to implement so I think duplicated test in `UserService` can be removed for now as they are supposed to be internal subcomponents of `Service`.

File: app_server/test/controllers/UserController.js (hunk 6 above)
Comment (ndt93) [inline app_server/test/controllers/UserController.js]: Some terminology used by the validation library `Joi`. Can just leave it there I suppose.

File: app_server/test/controllers/UserController.js (hunk 9 above)
Comment (yj-soh) [inline app_server/test/controllers/UserController.js]: Should the `function () {}` part be removed, then?
I'm not entirely sure if I'm looking up [the correct documentation](http://hapijs.com/api#serverstopoptions-callback), but the callback appears to be optional, so it may be better to remove it if it's not used.

File: app_server/test/services/UserService.js (hunk 10 above)
Comment (yj-soh) [inline app_server/test/services/UserService.js]: Yeah, that's what I thought too, but I'm not quite sure how to even implement that check.
Maybe leave some sort of comment for future reference regardless of whichever method is used, I guess.

File: app_server/test/controllers/UserController.js (hunk 8 above)
Comment (ndt93) [inline app_server/test/controllers/UserController.js]: Can use `underscore` for this actually.

File: app_server/test/controllers/UserController.js (hunk 9 above)
Comment (ndt93) [inline app_server/test/controllers/UserController.js]: Ah ok. That's cleaner then.

Comment (ndt93) [inline app_server/test/controllers/UserController.js]: Hmm seems like the docs is outdated. The callback is required.

**Comment by ndt93** [comment]
@yj-soh Fixed the mentioned issues. You can check this again.

**Comment by yj-soh** [comment]
If you're done with my comments and stuff, go ahead and merge~
Doubt there's anything else for me to point out (assuming it works fine).
